src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl

```diff
@@ -26,11 +26,10 @@ Models.components(model::Nonhydrostatic2DModel) = (
     vertical_diffusion = model.vertical_diffusion,
 )
 
-function Models.default_initial_conditions(model::Nonhydrostatic2DModel)
+function Models.default_initial_conditions(model::Nonhydrostatic2DModel, space_center, space_face)
     # we need to provide default initial conditions for the model, because the ode solver
     # requires inital conditions when getting instantiated, but we also want to support the `set!` function
     # interface for initialization and re-initialization.
-    space_center, space_face = Domains.make_function_space(model.domain)
```

Comment:
Is there a reason why you remove this line and make `space_center` and `space_face` input args? I prefer this old way because it looks simpler.

Reply:
Yes. They can only be initialised once per simulation. Otherwise you get an error `Broacasted spaces are the same ClimaCore.Spaces type but not the same instance` that comes from `ClimaCore` `Fields/broadcast.jl` file. 

I think it makes sense that you would want to only create one instance of space per simulation. Otherwise it would be hard to guarantee that they are the same between different variables. 

One option would be to initialise both the state and cache in one function. Then they could be hidden back inside?

Reply:
I think we'd need to remove and handle the function space creation somewhere else in the end, but we could break it down like you are proposing here :)